File: src/lua-binding/binding-helpers/src/binding-helpers.ts
module BindingHelpers {
	/**
	 * A wrapper object passed into the runtime to bind APIs.
	 */
	export type DataWrapper<T> = {
		/**
		 * Getter function passed in from the API.
		 * @returns A value provided by the API.
		 */
		get?: () => T;
		/**
		 * Setter function passed in from the API.
		 * @param value A value to set in the API.
		 */
		set?: (value: T) => void;
	};

	export type DataWrapperReadOnly<T> = Required<Pick<DataWrapper<T>, 'get'>>;
	export type DataWrapperSetOnly<T> = Required<Pick<DataWrapper<T>, 'set'>>;
	export type DataWrapperReadSet<T> = Required<DataWrapper<T>>;
	export type DataWrapperOptions<T> =
		| DataWrapperReadOnly<T>
		| DataWrapperSetOnly<T>
		| DataWrapperReadSet<T>;

	export type BindWrapperTargetResult<
		TargetType extends Object,
		Key extends string,
		Wrapper extends DataWrapper<any>
	> = TargetType & {
		[K in Key]: Wrapper extends DataWrapper<infer T> ? T : never;
	};

	/**
	 *
	 * @param key Key on destination to bind the wrapper to.
	 * @param wrapper Data wrapper provided by the API to bind to the destination.
	 * @param destination Destination to bind the wrapper on. Defaults to the module exports.
	 */
	export function bindWrapper<
		T,
		K extends string,
		TargetType extends Object,
		Wrapper extends DataWrapper<T>
	>(
		key: K,
		wrapper: DataWrapper<T>,
		destination: TargetType = module.exports
	) {
		Object.defineProperty(destination, key, wrapper);
	}

	export interface ExternalArrayAccessor<T> {
		getIndex: (index: number) => T;
		setIndex: (index: number, value: T) => void;
		push: (value: T) => number;
		pop: () => T;
		insert: (index: number, value: T) => number;
		length: () => number;
	}

	export class ArrayAccessor<T> {
		public static *getArrayIterator<T>(array: ArrayAccessor<T>) {
			for (let i = 0; i < array.length; i++) {
				yield array.getIndex(i);
			}
		}

		public [Symbol.iterator] = () => ArrayAccessor.getArrayIterator(this);

		public constructor(public accessor: ExternalArrayAccessor<T>) {
			this.getIndex = this.accessor.getIndex;
			this.setIndex = this.accessor.setIndex;
			this.push = this.accessor.push;
			this.pop = this.accessor.pop;
			this.insert = this.accessor.insert;
		}

		public getIndex: (index: number) => T;
		public setIndex: (index: number, value: T) => void;
		public push: (value: T) => number;
		public pop: () => T;
		public insert: (index: number, value: T) => number;

		public get length() {
			return this.accessor.length();
		}

		public indexOf(
			callback: (
				value: T,
				index: number,
				array: ArrayAccessor<T>
			) => boolean
		) {
			let i = 0;
			for (const value of this) {
				if (callback(value, i, this)) {
					break;
				}
				i++;
			}

			return i;
		}

		public find(
			callback: (
				value: T,
				index: number,
				array: ArrayAccessor<T>
			) => boolean
		) {
			return this.accessor.getIndex(this.indexOf(callback));
		}
	}

	export function bindArrayWrapper<T>(source: ExternalArrayAccessor<T>) {
		return new ArrayAccessor<T>(source);
	}

	type KeysMatching<T, V> = {
		[K in keyof T]-?: T[K] extends V ? K : never;
	}[keyof T];

	type ReadOnlyMembers<Members extends Record<string, DataWrapper<any>>> =
		Exclude<
			KeysMatching<Members, DataWrapperReadOnly<any>>,
			KeysMatching<Members, DataWrapperSetOnly<any>>
		>;
	type SetOnlyMembers<Members extends Record<string, DataWrapper<any>>> =
		Exclude<
			KeysMatching<Members, DataWrapperSetOnly<any>>,
			KeysMatching<Members, DataWrapperReadOnly<any>>
		>;
	type SetableMembers<Members extends Record<string, DataWrapper<any>>> =
		KeysMatching<Members, DataWrapperSetOnly<any>>;

	export type ClassPrototype<
		Members extends Record<string, DataWrapper<any>>
	> = {
		readonly [K in ReadOnlyMembers<Members>]: Members[K] extends DataWrapperReadOnly<
			infer T
		>
			? T
			: never;
	} & {
		[K in SetableMembers<Members>]: Members[K] extends DataWrapperSetOnly<
			infer T
		>
			? T
			: never;
	};

	export type ClassConstructor<
		ConstructorParameters extends unknown[],
		MemberFields extends Record<string, DataWrapper<any>>
	> = (...params: ConstructorParameters) => MemberFields;

	export type Class<
		ConstructorParameters extends unknown[],
		MemberFields extends Record<string, any>,
		StaticFields extends Record<string, any>,
		StaticFieldNames extends string = StaticFields extends Record<
			infer N,
			any
		>
			? N
			: never
	> = {
		new (...params: ConstructorParameters): ClassPrototype<MemberFields>;
	} & {
		[K in StaticFieldNames]: BindWrapperTargetResult<
			{},
			K,
			StaticFields[K]
		>[K];
	};

	class InitialEmptyConstructor {}

	class BindAsSrc<
		MemberFields extends Record<string, DataWrapperOptions<any>>
	> {
		constructor(public source: MemberFields) {}
	}

	export class APIClass {
		constructor() {}
	}

	/**
	 *
	 * @param className The name of the generated class
	 * @param constructor The function which generates the DataWrapper to bind as the class instance
	 * @param staticFields Static fields which are bound to the actual class object.
	 * @returns An object containing the generatedClass and a bindAs function which can be used by the
	 * 	C++ calling function to bind an object as an instance of the class.
	 */
	export function makeClass<
		ConstructorParameters extends unknown[],
		MemberFields extends Record<string, DataWrapperOptions<any>>,
		StaticFields extends Record<string, DataWrapperOptions<any>>,
		ClassName extends string
	>(
		className: ClassName,
		constructor: ClassConstructor<ConstructorParameters, MemberFields>,
		staticFields: StaticFields
	): {
		generatedClass: Class<
			ConstructorParameters,
			MemberFields,
			StaticFields
		>;
		bindAs: (sourceObject: MemberFields) => ClassPrototype<MemberFields>;
	} {
		/*
		Trick to get the class name to match what was passed into the function.
		This way, `class.name` will return that value.
	*/
		const classContainer = {
			[className]: class extends APIClass {
				constructor(
					...params:
						| ConstructorParameters
						| [InitialEmptyConstructor]
						| [BindAsSrc<MemberFields>]
				) {
					super();
					/*
					Trick to create an object that is an instance of the class,
					that way we can bind all of the member fields dynamically
					onto the object. Since the object was created with an actual
					constructor, it is an actual instance of the class, so the
					`instanceof` operator will still work.

					This is a workaround since TSTL does not support `Object.setPrototypeOf`.
				*/
					if (params[0] instanceof InitialEmptyConstructor) {
						return;
					}
					let constructorResult: MemberFields;
					if (params[0] instanceof BindAsSrc) {
						constructorResult = params[0].source;
					} else {
						constructorResult = constructor(
							...(params as ConstructorParameters)
						);
					}

					const constructed = new classContainer[className](
						new InitialEmptyConstructor()
					);

					for (const [key, value] of Object.entries(
						constructorResult
					)) {
						bindWrapper(key, value, constructed);
					}

					return constructed as ClassPrototype<MemberFields>;
				}
			}
		};

		for (const [key, value] of Object.entries(staticFields)) {
			bindWrapper(key, value, classContainer[className] as any);
		}

		const generatedClass = classContainer[className] as Class<
			ConstructorParameters,
			MemberFields,
			StaticFields
		>;

		const generatedBindAsClass: {
			new (source: BindAsSrc<MemberFields>): ClassPrototype<MemberFields>;
		} = generatedClass as unknown as any;

		return {
			generatedClass,
			bindAs: (sourceObject: MemberFields) =>
				new generatedBindAsClass(
					new BindAsSrc<MemberFields>(sourceObject)
				)
		};
	}
}

const moduleWrapper = { BindingHelpers };
type ModuleName = keyof typeof moduleWrapper;
export const bindingHelpersModuleName: ModuleName = Object.keys(
	moduleWrapper
)[0] as ModuleName;
export const moduleResolutionPath = 'binding-helpers';

export default BindingHelpers;

File: src/lua-binding/binding-helpers/test/src/binding-helpers.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectKeys = ____lualib.__TS__ObjectKeys
local __TS__ObjectDefineProperty = ____lualib.__TS__ObjectDefineProperty
local __TS__Class = ____lualib.__TS__Class
local __TS__Symbol = ____lualib.__TS__Symbol
local Symbol = ____lualib.Symbol
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__Generator = ____lualib.__TS__Generator
local __TS__Iterator = ____lualib.__TS__Iterator
local __TS__New = ____lualib.__TS__New
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__InstanceOf = ____lualib.__TS__InstanceOf
local __TS__Unpack = ____lualib.__TS__Unpack
local __TS__ObjectEntries = ____lualib.__TS__ObjectEntries
local ____exports = {}
local BindingHelpers
local moduleWrapper = {BindingHelpers = BindingHelpers}
____exports.bindingHelpersModuleName = __TS__ObjectKeys(moduleWrapper)[1]
____exports.moduleResolutionPath = "binding-helpers"
BindingHelpers = {}
do
    ---
    -- @param key Key on destination to bind the wrapper to.
    -- @param wrapper Data wrapper provided by the API to bind to the destination.
    -- @param destination Destination to bind the wrapper on. Defaults to the module exports.
    function BindingHelpers.bindWrapper(self, key, wrapper, destination)
        if destination == nil then
            destination = module.exports
        end
        __TS__ObjectDefineProperty(destination, key, wrapper)
    end
    BindingHelpers.ArrayAccessor = __TS__Class()
    local ArrayAccessor = BindingHelpers.ArrayAccessor
    ArrayAccessor.name = "ArrayAccessor"
    function ArrayAccessor.prototype.____constructor(self, accessor)
        self.accessor = accessor
        self[Symbol.iterator] = function() return BindingHelpers.ArrayAccessor:getArrayIterator(self) end
        self.getIndex = self.accessor.getIndex
        self.setIndex = self.accessor.setIndex
        self.push = self.accessor.push
        self.pop = self.accessor.pop
        self.insert = self.accessor.insert
    end
    __TS__SetDescriptor(
        ArrayAccessor.prototype,
        "length",
        {get = function(self)
            return self.accessor:length()
        end},
        true
    )
    ArrayAccessor.getArrayIterator = __TS__Generator(function(self, array)
        do
            local i = 0
            while i < array.length do
                coroutine.yield(array:getIndex(i))
                i = i + 1
            end
        end
    end)
    function ArrayAccessor.prototype.indexOf(self, callback)
        local i = 0
        for ____, value in __TS__Iterator(self) do
            if callback(nil, value, i, self) then
                break
            end
            i = i + 1
        end
        return i
    end
    function ArrayAccessor.prototype.find(self, callback)
        return self.accessor:getIndex(self:indexOf(callback))
    end
    function BindingHelpers.bindArrayWrapper(self, source)
        return __TS__New(BindingHelpers.ArrayAccessor, source)
    end
    local InitialEmptyConstructor = __TS__Class()
    InitialEmptyConstructor.name = "InitialEmptyConstructor"
    function InitialEmptyConstructor.prototype.____constructor(self)
    end
    local BindAsSrc = __TS__Class()
    BindAsSrc.name = "BindAsSrc"
    function BindAsSrc.prototype.____constructor(self, source)
        self.source = source
    end
    BindingHelpers.APIClass = __TS__Class()
    local APIClass = BindingHelpers.APIClass
    APIClass.name = "APIClass"
    function APIClass.prototype.____constructor(self)
    end
    ---
    -- @param className The name of the generated class
    -- @param constructor The function which generates the DataWrapper to bind as the class instance
    -- @param staticFields Static fields which are bound to the actual class object.
    -- @returns An object containing the generatedClass and a bindAs function which can be used by the
    -- C++ calling function to bind an object as an instance of the class.
    function BindingHelpers.makeClass(self, className, constructor, staticFields)
        local classContainer
        local ____className_1 = className
        local ____class_0 = __TS__Class()
        ____class_0.name = ____class_0.name
        __TS__ClassExtends(____class_0, BindingHelpers.APIClass)
        function ____class_0.prototype.____constructor(self, ...)
            local params = {...}
            ____class_0.____super.prototype.____constructor(self)
            if __TS__InstanceOf(params[1], InitialEmptyConstructor) then
                return
            end
            local constructorResult
            if __TS__InstanceOf(params[1], BindAsSrc) then
                constructorResult = params[1].source
            else
                constructorResult = constructor(
                    nil,
                    __TS__Unpack(params)
                )
            end
            local constructed = __TS__New(
                classContainer[className],
                __TS__New(InitialEmptyConstructor)
            )
            for ____, ____value in ipairs(__TS__ObjectEntries(constructorResult)) do
                local key = ____value[1]
                local value = ____value[2]
                BindingHelpers.bindWrapper(nil, key, value, constructed)
            end
            return constructed
        end
        classContainer = {[____className_1] = ____class_0}
        for ____, ____value in ipairs(__TS__ObjectEntries(staticFields)) do
            local key = ____value[1]
            local value = ____value[2]
            BindingHelpers.bindWrapper(nil, key, value, classContainer[className])
        end
        local generatedClass = classContainer[className]
        local generatedBindAsClass = generatedClass
        return {
            generatedClass = generatedClass,
            bindAs = function(____, sourceObject) return __TS__New(
                generatedBindAsClass,
                __TS__New(BindAsSrc, sourceObject)
            ) end
        }
    end
end
____exports.default = BindingHelpers
return ____exports
